mindmeld/active_learning/classifiers.py

```diff
@@ -0,0 +1,418 @@
+        self.intent2idx, self.idx2intent, self.domain_indices = self._get_mappings()
+
+    def _get_mappings(self):
+        """Get mappings of intents to indices and the indices that map to each domain.
+
+        Returns:
+            intent2idx (Dict): Maps intents to indices
+            idx2intent (Dict): Maps indices to intents
+            domain_indices (Dict): Maps domains to a tuple containing the start and
+                ending indexes of intents with the given domain.
+        """
+        idx = 0
+        intent2idx, idx2intent, domain_indices = {}, {}, {}
+        for domain in sorted(os.listdir(os.path.join(self.app_path, "domains"))):
+            start_idx = idx
+            for intent in sorted(
+                os.listdir(os.path.join(self.app_path, "domains", domain))
+            ):
+                intent2idx[f"{domain}|{intent}"] = idx
+                idx2intent[idx] = f"{domain}|{intent}"
+                idx += 1
+            end_idx = idx - 1
+            domain_indices[domain] = (start_idx, end_idx)
+        return intent2idx, idx2intent, domain_indices
+
+    @abstractmethod
+    def train(self):
+        raise NotImplementedError(
+            "Subclasses must implement their classifier's fit method."
+        )
+
+
+class MindMeldALClassifier(ALClassifier):
+    """Active Learning classifier that uses MindMeld classifiers internally.
+    Handles the training of MindMeld components (Domain or Intent classifiers)
+    and collecting performance statistics (eval_stats)."""
+
+    def __init__(self, app_path: str, training_level: str, n_classifiers: int):
+        """
+        Args:
+            app_path (str): Path to MindMeld application
+            training_level (str): The hierarchy level to train ("domain" or "intent")
+            n_classifiers (int): Number of classifiers to be used by multi-model strategies.
+        """
+        super().__init__(app_path=app_path, training_level=training_level)
+        self.nlp = NaturalLanguageProcessor(self.app_path)
+        self.n_classifiers = n_classifiers
+
+    @staticmethod
+    def _get_probs(
+        classifier: Classifier, queries: List[ProcessedQuery], nlp_component_to_id: Dict
+    ):
+        """Get the probability distribution for for a query across domains or intents.
+
+        Args:
+            classifier (MindMeld Classifer): Domain or Intent Classifier
+            queries (List of ProcessedQuery): List of MindMeld queries
+            nlp_component_to_id (Dict): Dictionary mapping domain or intent names to vector index
+                positions.
+
+        Returns:
+            prob_vector (List[List]]): Probability distribution vectors for given queries.
+        """
+        queries_prob_vectors = []
+        if queries:
+            classifier_eval = classifier.evaluate(queries=queries)
+            for x in classifier_eval.results:
+                query_prob_vector = [None] * len(nlp_component_to_id)
```

Comment:
If we're calling it a "query probability vector", we should assign it to a zero probability vector rather than a list of Nones.